# Supervised Learning Model Evaluation Lab

Complete the exercises below to solidify your knowledge and understanding of supervised learning model evaluation.

In [1]:
import pandas as pd

## Regression Model Evaluation

In [43]:
from sklearn.datasets import load_boston

data = load_boston()

X_boston = pd.DataFrame(data["data"], columns=data["feature_names"])
y_boston = pd.DataFrame(data["target"], columns=['MEDV'])

data = pd.concat([X, y], axis=1)

## 1. Split this data set into training (80%) and testing (20%) sets.

The `MEDV` field represents the median value of owner-occupied homes (in $1000's) and is the target variable that we will want to predict.

In [3]:
from sklearn.model_selection import train_test_split

In [44]:
X_train_boston, X_test_boston, y_train_boston, y_test_boston = train_test_split( X_boston, 
                                                    y_boston, 
                                                    test_size=0.20, 
                                                    random_state=123)

## 2. Train a `LinearRegression` model on this data set and generate predictions on both the training and the testing set.

In [5]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [46]:
lr = LinearRegression()

lr.fit(X_train_boston, y_train_boston)

y_train_pred = lr.predict(X_train_boston)

y_test_pred = lr.predict(X_test_boston)

## 3. Calculate and print R-squared for both the training and the testing set.

In [47]:
r2 = [round(r2_score(y_train_boston, y_train_pred),3),round(r2_score(y_test_boston, y_test_pred),3) ]
r2

[0.756, 0.659]

## 4. Calculate and print mean squared error for both the training and the testing set.

In [48]:
from sklearn.metrics import mean_squared_error as mse

In [49]:
mean_square = [round(mse(y_train_boston, y_train_pred),3), round(mse(y_test_boston, y_test_pred),3)]
mean_square

[20.677, 28.192]

## 5. Calculate and print mean absolute error for both the training and the testing set.

In [10]:
from sklearn.metrics import mean_absolute_error as mae

In [50]:
abs_error = [round(mae(y_train_boston, y_train_pred),3), round(mae(y_test_boston, y_test_pred),3)]

abs_error

[3.175, 3.66]

#### The r2 score for the training set is greater than the score for the testing set, and the MSE is lower also lower in the former. If these were independent models we could say that the first one is better. 
#### In this case they belong to the same set and model, with just differentiating between trainin and test, it's reasonable to see that there are predicitions done on the training set are more accurate that those of the testing set, as the model has been trained with those same values. Also, the mean absolute error is lower in the training set. 
#### Having said this, the model is not at all bad, with an r2 of 65.9%, a MSE of 28,19 and an MAE of 3.66. 

## Classification Model Evaluation

In [51]:
from sklearn.datasets import load_iris

data = load_iris()

X_iris = pd.DataFrame(data["data"], columns=data["feature_names"])
y_iris = pd.DataFrame(data["target"], columns=["class"])

data = pd.concat([X, y], axis=1)

In [52]:
y_iris["class"].value_counts()
# There are three classes, so we should use a multiclass aproach 

2    50
1    50
0    50
Name: class, dtype: int64

## 6. Split this data set into training (80%) and testing (20%) sets.

The `class` field represents the type of flower and is the target variable that we will want to predict.

In [53]:
X_train_iris, X_test_iris, y_train_iris, y_test_iris = train_test_split( X_iris, 
                                                    y_iris, 
                                                    test_size=0.20,
                                                   random_state=150)

## 7. Train a `LogisticRegression` model on this data set and generate predictions on both the training and the testing set.

In [157]:
from sklearn.linear_model import LogisticRegression

In [155]:
lg = LogisticRegression(max_iter = 1000)

lg.fit(X_train_iris, y_train_iris.values.ravel()) # A warning appeared about the size of the y_train and suggested using .raven()

y_train_pred = lg.predict(X_train_iris)

y_test_pred = lg.predict(X_test_iris)


## 8. Calculate and print the accuracy score for both the training and the testing set.

In [30]:
from sklearn.metrics import accuracy_score

In [55]:
training_accuracy = accuracy_score(
    y_true=y_train_iris,
    y_pred=y_train_pred
)

testing_accuracy = accuracy_score(
    y_true=y_test_iris,
    y_pred=y_test_pred
)

results = [round(training_accuracy, 3), round(testing_accuracy,3)]

results

[0.983, 0.967]

## 9. Calculate and print the balanced accuracy score for both the training and the testing set.

In [32]:
from sklearn.metrics import balanced_accuracy_score as bas

In [56]:
balanced_accuracy = [round(bas(y_train_iris, y_train_pred),3), round(bas(y_test_iris, y_test_pred),3)]
balanced_accuracy

[0.982, 0.972]

## 10. Calculate and print the precision score for both the training and the testing set.

In [34]:
from sklearn.metrics import precision_score as ps

In [59]:
precision_score = [round(ps(y_train_iris, y_train_pred, average = 'weighted'),3), 
                   round(ps(y_test_iris, y_test_pred, average = 'weighted'),3)]
precision_score

[0.984, 0.97]

## 11. Calculate and print the recall score for both the training and the testing set.

In [36]:
from sklearn.metrics import recall_score as rs

In [60]:
recall_score = [round(rs(y_train_iris, y_train_pred, average = 'weighted'), 3), 
                round(rs(y_test_iris, y_test_pred, average = 'weighted'),3)]
recall_score

[0.983, 0.967]

## 12. Calculate and print the F1 score for both the training and the testing set.

In [38]:
from sklearn.metrics import f1_score as f1

In [61]:
f_one = [round(f1(y_train_iris, y_train_pred, average = "weighted"),3),
         round(f1(y_test_iris, y_test_pred, average = 'weighted'),3)]
f_one

[0.983, 0.967]

## 13. Generate confusion matrices for both the training and the testing set.

In [40]:
from sklearn.metrics import confusion_matrix

In [63]:
confusion_matrix(y_train_iris, y_train_pred)

array([[41,  0,  0],
       [ 0, 36,  2],
       [ 0,  0, 41]], dtype=int64)

In [62]:
confusion_matrix(y_test_iris, y_test_pred)

array([[ 9,  0,  0],
       [ 0, 11,  1],
       [ 0,  0,  9]], dtype=int64)

#### When we have multiclass the confusion matrix changes a little bit. Instead os showing the True and False + -, the upper side represents the true values, while the left side represents the predicted valuesa; and in each row we have the different classes (0,1,2)

#### In the test matrix we can see that the model's predictions are all correct, except for a false positive of a value predicted as 1 when it was really a 2. 

#### In the training set the matrix shows a similar result, except for two value that have been predicted as 1 when they were 2

## Bonus: For each of the data sets in this lab, try training with some of the other models you have learned about, recalculate the evaluation metrics, and compare to determine which models perform best on each data set.

### Decision tree for Boston data

In [79]:
from sklearn.tree import DecisionTreeRegressor

In [122]:
gs = GridSearchCV(
    estimator=DecisionTreeRegressor(),
    param_grid={
        "max_depth": [11,12,13],
        "min_samples_split": [9,10,11,12],
        "max_features": [8,9]
    },
    cv=5,
    verbose=3,
    return_train_score=True
)

In [ ]:
gs.fit(X_train_boston, y_train_boston)

In [125]:
print(gs.best_estimator_)
print(gs.best_score_)

# After a series of tries we can say that the best model found is one with a max_depth of 12, a max_features of 9 and min_sample
# of 11. However, there is probably a better model 

DecisionTreeRegressor(max_depth=12, max_features=9, min_samples_split=11)
0.81101920940787


In [128]:
best_dt = DecisionTreeRegressor(max_depth = 12, max_features = 9, min_samples_split = 11)

In [129]:
best_dt.fit(X_train_boston, y_train_boston)

DecisionTreeRegressor(max_depth=12, max_features=9, min_samples_split=11)

In [133]:
y_pred = best_dt.predict(X_test_boston)

In [136]:
round(mse(y_test_boston, y_pred),3)

17.627

In [137]:
round(r2_score(y_test_boston, y_pred),3)

0.787

#### With the decision tree model we have been able to improve the model for the Boston data. The original r2 and MSE of the linear regression was 0.659 and 28.192, respectively; while the same parametres for the Decision Tree are 0.788 and 17.627

### KNeighbors for Iris dataset

In [138]:
from sklearn.neighbors import KNeighborsClassifier

In [142]:
n_neighbors = range(1,20)

model = KNeighborsClassifier()
parameter_space = {'n_neighbors': n_neighbors,
                   }

gs = GridSearchCV(model,
                           param_grid=parameter_space,
                           cv=5)

gs.fit(X_train_iris, y_train_iris.values.ravel())

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': range(1, 20)})

In [145]:
print(gs.best_estimator_)
print(gs.best_score_)

# The best model is the one with n_neighbors = 4

KNeighborsClassifier(n_neighbors=4)
0.975


In [162]:
kn = KNeighborsClassifier(n_neighbors = 4)

In [163]:
kn.fit(X_train_iris, y_train_iris.values.ravel())

KNeighborsClassifier(n_neighbors=4)

In [164]:
y_pred = kn.predict(X_test_iris)

In [165]:
print(round(mse(y_test_iris, y_pred),3))
print(round(r2_score(y_test_iris, y_pred),3))

0.033
0.944


In [166]:
print(round(mse(y_test_iris, y_test_pred),3))
print(round(r2_score(y_test_iris, y_test_pred),3))

0.033
0.944


#### We get the same results for the KNeighbors and the Logisitic model